packages/model/src/result.ts

```diff
@@ -38,21 +36,20 @@ class Failure<A, E> {
     this.error = error
   }
 
-  then = <B>(_f: (value: A) => Result<B, E>): Result<B, E> => fail(this.error)
+  chain = <B>(_f: (value: A) => Result<B, E>): Result<B, E> => fail(this.error)
   replace = <B>(_value: B): Result<B, E> => fail(this.error)
   map = <B>(_f: (value: A) => B): Result<B, E> => fail(this.error)
   mapError = <E1>(f: (error: E) => E1): Result<A, E1> => fail(f(this.error))
-  unwrap = (fallback: A): A => fallback
-  lazyUnwrap = (fallback: () => A): A => fallback()
+  recover = (fromError: (error: E) => A): A => fromError(this.error)
   or = (other: Result<A, E>): Result<A, E> => other
-  lazyOr = (other: () => Result<A, E>): Result<A, E> => other()
+  lazyOr = (other: (error: E) => Result<A, E>): Result<A, E> => other(this.error)
   match = <B>(_onOk: (value: A) => B, onFailure: (error: E) => B): B => onFailure(this.error)
 }
 
-export function ok<A, E>(value: A): Result<A, E> {
+export function ok<A, E>(value: A): Ok<A, E> {
```

Comment:
@edobrb why make this return the more specific value? Does this have any advantage?

Reply:
I like to return the narrowest type. I would expect to be able to write this:

```typescript
const result = ok(10)
console.log(result.value) //already narrowed
```

instead of:

```typescript
const result = ok(10)
if(result.isOk) {
  console.log(result.value)
}
```